### PR TITLE
[KEYCLOAK-11746] More unit tests: browser flow

### PR DESCRIPTION
**Pull request by fperot74**
Replaces cloudtrust/keycloak#18

### CODE REVIEW

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/FlowUtil.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            List<AuthenticationExecutionModel> execs = realm.getAuthenticationExecutions(currentFlow.getId());
+            if (execs == null) {
+                throw new FlowUtilException("Can't get executions of unknown flow " + currentFlow.getId());
+            }
+            executions = new ArrayList<>(execs);
+        }
+        return executions;
+    }
+
+    public FlowUtil removeExecution(int index) {
```
Comment (AlistairDoswald) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/FlowUtil.java:239]: Index is never used, instead index 0 is always removed

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/BrowserFlowTest.java [hunk 2]
```diff
@@ -190,10 +214,258 @@ public void userWithTwoAdditionalFactors() throws InterruptedException {
...
+    // In a form waiting for a username only, provides a username and check if password is requested in the following execution of the flow
+    private boolean needsPassword(String username) {
+        // provides username
+        loginUsernameOnlyPage.open();
+        loginUsernameOnlyPage.login(username);
+
+        return passwordPage.isCurrent();
+    }
+
+    // A conditional block without conditional block should automatically be disabled
```
Comment (AlistairDoswald) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/BrowserFlowTest.java]: A conditional _flow_ without conditional block should automatically be disabled

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/BrowserFlowTest.java [hunk 3]
```diff
@@ -190,10 +214,258 @@ public void userWithTwoAdditionalFactors() throws InterruptedException {
...
+                            }
+                            // Update the browser forms only with a UsernameForm
+                            subFlow.addAuthenticatorExecution(Requirement.REQUIRED, PasswordFormFactory.PROVIDER_ID);
+                        }))
+                .defineAsBrowserFlow()
+        );
+    }
+
+    // Configure a conditional block in a non-conditional sub-flow
+    // In such case, the whole sub-flow should be disabled
```
Comment (AlistairDoswald) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/BrowserFlowTest.java]: It should be "// In such case, the conditional block is not evaluated", no? And this is what this test shows

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/BrowserFlowTest.java [hunk 4]
```diff
@@ -190,10 +214,258 @@ public void userWithTwoAdditionalFactors() throws InterruptedException {
...
+                                        config -> config.getConfig().put("condUserRole", requiredRole))
+                                .addAuthenticatorExecution(Requirement.REQUIRED, OTPFormAuthenticatorFactory.PROVIDER_ID)
+                        )
+                )
+                .defineAsBrowserFlow()
+        );
+    }
+
+    @Test
+    public void testAlternativeSubFlowWithAlwaysValidatingExecutor() {
```
Comment (AlistairDoswald) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/BrowserFlowTest.java]: Is this test incomplete?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/FlowUtil.java [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        realm.setBrowserFlow(currentFlow);
+        return this;
+    }
+
+    public FlowUtil defineAsDirectGrantFlow() {
+        realm.setDirectGrantFlow(currentFlow);
+        return this;
+    }
+
+    public FlowUtil defineAsFlow() {
```
Comment (AlistairDoswald) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/FlowUtil.java]: Should be called "defineAsResetCredentialsFlow"

**Review by AlistairDoswald — CHANGES_REQUESTED** [review]
Some small naming and errors in the comments and one function which will not perform as expected.

I'm not sure what is happening in test testAlternativeSubFlowWithAlwaysValidatingExecutor... it doesn't seem to be testing much, and the flow doesn't seem to match what the name of the function describes. For this I suggest either removing the test for now, or accepting it, but in both cases correcting it in a 2nd PR.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/FlowUtil.java (hunk 1 above)
Comment (fperot74) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/FlowUtil.java]: You're right... I started working on it but had to switch on something else... I remove it for now.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/BrowserFlowTest.java (hunk 4 above)
Comment (fperot74) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/BrowserFlowTest.java]: Right... I started working on it but had to switch on something else...
I removing it.

**Review by AlistairDoswald — APPROVED** [review]

**Pull request merged**